### ISSUE TITLE
spawnProcess should reset signal handlers to default before exec()ing new executable

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4199]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#4199|
|Type|enhancement|
|Created|2010-01-05 16:38:03Z|
|Branch|https://github.com/twisted/twisted/tree/reset-signal-handlers-4199|

Python (and presumably applications) override certain signals, in particular SIGPIPE; spawnProcessed processes will have SIGPIPE in a state they don't necessarily expect. We should reset signal handlers to their default before calling exec().

Ticket [#4159](https://github.com/twisted/twisted/issues/4159) would allow this, but that still requires users to figure out they need to do this. It would be better to have good defaults and documented behavior.

See http://bugs.python.org/issue1652 for a detailed write-up of Python's perspective on the issue.

Attachments:

* [process.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/aef/aef2a867acf487ce94256c4e3b5ca98cebed8e0a/4eb421eb29f695708a54bee956d52909b4dfcffb.diff) (2321 bytes) - added by donal on 2010-01-11 00:40:09Z - Diff of process.py against trunk
* [testChildSignalHandling.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/aef/aef2a867acf487ce94256c4e3b5ca98cebed8e0a/625bc7d4dc4f656f36eee7edb6485b921a2f8469.diff) (2409 bytes) - added by donal on 2010-01-12 10:56:44Z - Diff of t.i.process.py and t.test.test_process.py
* [4199-3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/aef/aef2a867acf487ce94256c4e3b5ca98cebed8e0a/2eafeddaea8166636b88fb9b42e1e3389040dfde.diff) (4188 bytes) - added by PenguinOfDoom on 2010-01-17 10:54:15Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4199 4199
type__enhancement enhancement
reporter__itamarst itamarst
priority__normal normal
milestone__ 
branch__branches_reset_signal_handlers_4199 branches/reset-signal-handlers-4199
branch_author__donal__PenguinOfDoom__exarkun donal, PenguinOfDoom, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1262709483000000 1262709483000000
changetime__1265228476000000 1265228476000000
version__None None
owner__ 
cc__spiv cc__exarkun cc__glyph cc__PenguinOfDoom
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="donal's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|donal commented|
|-|-|

I just added a diff for process.py - what d'you think of this approach? 

If it's ok, do we need to also revise _BaseProcess.signalProcess to make use of the _BaseProcess.signals[] list?

I'm pretty new to this Twisted stuff (in case that's not already obvious) but I'm happy to have this assigned to me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to donal|
|-|-|

Thanks for the patch!  It's a big help. :)

For some reason I expected this to be done with a single call that reset everything at once.  I'm not sure why I had that expectation, though (I didn't even know about `sigvec` until just now, after reading various signal-related man pages for 5 or 10 minutes).  Harumph.

Anyway, a few minor comments:

  1. Rather than listing signal names, perhaps it would be okay to just iterate over `range(1, signal.NSIGS)`?  I haven't tried this out, but it seems like it would simplify the code a bit.
  1. According to sigaction(2), execve(2) resets the dispositions of _handled signals_ to the default.  It goes on to say that _the dispositions of ignored signals are left unchanged_.  To me, this sounds like only `SIG_IGN` handlers need to be reset to `SIG_DFL`.  Perhaps that means that rather than the `None` check and the `RuntimeError` handler, the code can just check to see if the handler is `SIG_IGN`.  Anything else either doesn't need to be re-set or will be re-set by the upcoming execve(2).
  1. The comment above the new loop in `_setupChild` says _before we fork_, but this code is actually run after the fork has happened (which is the correct thing to do - we don't want to reset the handlers in the parent, clearly).

Aside from that, the patch just needs to include a unit test.  This might not end up very "unity" since processes are so gross, but we need one all the same.  I can't really think of anything better than setting `SIG_IGN` for some signal, launching a child, signaling it, and then making sure it doesn't ignore the signal (somehow; ugh, this will be hard to do without a timeout or a race condition.  Of the two, timeouts are preferred, of course).

Thanks again for the patch!  When you have an updated one ready, make sure to add _review_ to the keywords list to make sure someone notices there's something to look at.

**Comment by twisted-trac** [comment]
|<img alt="donal's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|donal commented|
|-|-|

Hey that's much better - thanks for the review. I have a patch - I'll try to write a unit test tomorrow. I'm thinking it will:

```
- configure SIGINT to be ignored
- start a child with something like '/bin/sleep 10'
- send SIGINT to the child. (It should quit)
- reset the SIGINT handler
```

Sound ok? Cheers mate.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|spiv commented|
|-|-|

Couldn't use you use pipes to avoid race conditions and timeouts?

e.g. have the child process run `sys.stdout.write('ready\n')`, and have the parent wait for that before sending SIGINT?  That guarantees that the child has `fork`ed and `exec`ed, and so you know how it will react to SIGINT.

Actually, why send a signal at all?  Surely we can trust `signal.getsignal` to function correctly.  So couldn't you:

  * set SIGPIPE to have a custom handler in parent
  * spawn a child that does "`import signal; print signal.getsignal(signal.SIGPIPE) == signal.SIG_IGN`" (the default behaviour for SIGPIPE is to ignore the signal)
  * `assertEquals("True\n", childStdout)` in the parent.

Regardless, SIGINT is probably a bad choice for this test, because I suspect Python will unconditionally install its own SIGINT handler when it starts the interpreter.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Alas, we can't rely on `signal.getsignal`.  This just returns what Python thinks the signal handler is, not what the signal handler really is.  Also, `SIGPIPE` is a bad choice, because Python also changes the handler for that signal (to `SIG_IGN`) when it starts up.

As far as timeouts go, I was mainly thinking that if the implementation doesn't reset things properly, then the test is going to fail by timing out after two minutes.  Not the end of the world, but tests like that are no fun.

**Comment by twisted-trac** [comment]
|<img alt="donal's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|donal commented|
|-|-|

I'd figured that the child process we start cannot be a Python script, because: Our code will dutifully reset all the signal handlers to their defaults, and then exec(). If we then start up a Python interpreter, that interpreter will straightaway install its own signal handlers, so we're back where we started.

So the ultimate child process for the test case cannot be a Python process...?

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|spiv commented|
|-|-|

Hmm, my reading of Python's source is that it doesn't set all signal handlers, just SIGPIPE, SIGXFZ, SIGXFSZ and SIGINT.

So, to pick one at random, how about SIGUSR1?

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|spiv commented|
|-|-|

FWIW:

```
$ python -c "import os, sys, signal; print signal.getsignal(signal.SIGCHLD); 
signal.signal(signal.SIGCHLD, signal.SIG_IGN); print signal.getsignal(signal.SIGCHLD); 
os.execl(sys.executable, 'python', '-c', 'import signal; print signal.getsignal(signal.SIGCHLD)')"
0
1
1
```

```
$ python -c "import os, sys, signal; print signal.getsignal(signal.SIGWINCH); 
signal.signal(signal.SIGWINCH, signal.SIG_IGN); print signal.getsignal(signal.SIGWINCH); 
os.execl(sys.executable, 'python', '-c', 'import signal; print signal.getsignal(signal.SIGWINCH)')"
0
1
1
```

```
$ python -c "import os, sys, signal; print signal.getsignal(signal.SIGUSR1); 
signal.signal(signal.SIGUSR1, signal.SIG_IGN); print signal.getsignal(signal.SIGUSR1); 
os.execl(sys.executable, 'python', '-c', 'import signal; print signal.getsignal(signal.SIGUSR1)')"
0
1
1
```

So it appears there are some signal handler changes that are still visible in a process, even after exec'ing Python.

That said, it'd certainly be easier to test this stuff with confidence with a C program.

**Comment by twisted-trac** [comment]
|<img alt="donal's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|donal commented|
|-|-|

Fresh patch attached!

Hey Spiv - just about to upload a patch and dash off but I'll check your input more closely tomorrow. I've been looking at the sigign line in '/proc/PID#/status' hoping for a row of zeroes.

A regular process:
```
SigIgn: 0000000000000000
```

A python process:

```
SigIgn: 0000000001001000
```

For a python process there will normally be a couple of 1s in there. I worked out which signal each zero represents back when I was first working on this, but I'm not sure if I can find my notes on that. I can't find it documented anywhere on the internets.

The unit test starts a child process ('/bin/sleep') and sends it an interrupt. The process quits (as it does if interrupted after having been started from a shell). Without the fresh code in twisted/internet/process.py, the process just ignores the signal.

Cheers guys - glad this is moving :)

**Comment by twisted-trac** [comment]
|<img alt="donal's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|donal commented|
|-|-|

BTW - If you're doing cool process stuff with Python, the PSI module might be interesting
http://psychofx.com/psi/

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|PenguinOfDoom commented|
|-|-|

After this change, "trial --until-fail twisted/test/test_process.py twisted/test/test_tcp.py" crashes with a SIGPIPE.

What happens is, MockProcessTestCase fakes up os.fork, so the new signal-restoring code in _setupChild executes in the parent trial process. test_tcp happens to trigger SIGPIPE, with deadly results.

Also, the bug applies to PTYProcess as well, as neglected as it may be.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [28063]) Branching to 'reset-signal-handlers-4199'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [28064]) Apply 4199-3.diff

refs [#4199](https://github.com/twisted/twisted/issues/4199)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

We should try to avoid increasing the amount of duplication between `Process` and `PTYProcess`.  Also, the mock stuff bothers me.  I'm going to see if I can do anything about it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I didn't manage to do anything about the mock stuff.  I think that at some point we should, but maybe we can live with the slight expansion in that area that this branch adds for now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve set owner to @exarkun|
|-|-|

It looks ok, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set status to `closed`|
|-|-|

(In [28113]) Merge reset-signal-handlers-4199

Author: donal, PenguinOfDoom, exarkun
Reviewer: therve
Fixes: [#4199](https://github.com/twisted/twisted/issues/4199)

Reset the disposition of ignored signals in the child process created by
`reactor.spawnProcess` before executing the specified executable.  This
gives children a more normal environment (ie, one where SIGPIPE and,
potentially, other signals are in the unusual state of being ignored).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
